### ISSUE TITLE
false positive: inferred tuple[int] | tuple[None] vs tuple[int | None]

### ISSUE THREAD

**Issue by kurtbrose** [microsoft/pyright#9358]
**Describe the bug**
`tuple[A | B]` vs `tuple[A] | tuple[B]` is inferred differently for different constants, leading to false positive type errors

(I'm not sure if this is really a bug or not -- maybe the search space of equivalent type unions is too huge in the general case.  Inuititvely, it seems that a simple internal rule like always try to merge unions or always try to split unions would at least result in arriving at consistent inferred types.)

**Code or Screenshots**
https://pyright-play.net/?code=MQAg7glgLgFiB2B7EBnADgUwMYQGYQwBMQoBPTAGhJgwCcMByFEAQ3hDtsVoCgA3FgBtmAXhABvABQBGCgEoAXCFkhJAOUTwM8pdIC%2BIEKAHCQAawylmLeiAjxcnIq2ZQArmkEYA2vagBdEAAfEg8vbw0tfx5vExRvSQAPeUDcbhBEu3YZKkiMOUCjDLtme0daemIWVzCfP2CQPOjQTnSIZgAiAEFaAHM3AFsMeCgQRFwScgxQzx8AGWg6IW9pQJCmkCw2JFGAI2nqlAherWIoZDQbFiGoOnNLMYn3Wd8RtZnwpo6eHlA8cGmAzYbiEglIqEwOFw4Ng00IECwowsViooJKAMEgn4QhQACYlPDEd5nuF6utNBh-FQ-IExFJZIplFR1BSdMo9DE4riEskCiA0rRivZVCo8gUeEA

```python
# with no specified type, there's an error
vals = {(1,): 1, (None,): 1}  # vals keys are inferred as tuple[int] | tuple[None]
[vals[(x,)] for x in (1, None)]  # x is inferred as tuple[int | None]
# error is "Argument of type tuple[Literal[1] | None] cannot be assigned to parameter key of tuple[int] | tuple[None]"

# if we manually specify the dict key type, all is well
vals2: dict[tuple[int | None], int] = {(1,): 1, (None,): 1}
[vals2[(x,)] for x in (1, None)]
```

**VS Code extension or command-line**
This came up for me in VS Code at first, but hopefully the playground link makes it easy to reproduce.

**Comment by erictraut** [comment]
Pyright is working as intended here. This isn't a bug.

For details about how pyright infers the type of dictionary and tuple expressions, refer to [this documentation](https://microsoft.github.io/pyright/#/type-inference?id=tuple-expressions).

Regardless of the standard inference rules employed by a static type checker, there will always be cases where you may want a different type to be inferred. In such cases, you need to provided explicit type annotations, like you've done above for `vals2`.

**Comment by kurtbrose** [comment]
Thank you for the fast response.

I'm not sure if I'm misunderstanding, but I just wanted to follow up in case you have time and I'm really close to an "ah-ha" moment.

There's two things I'd like to understand: why are the types inferred differently?  why aren't `tuple[A | B]` and `tuple[A] | tuple[B]` equivalent?

Trying to follow the docs and trace the types:
* `strictDictionaryInference` would lead to `Union(keys)` combining `tuple[int]` and `tuple[None]` in the `dict` literal
* (I'm speculating) the list comprehension `[(x,) for x in (1, None)]` sees `x` as `Union(int, None)` and just wraps that in `tuple[]`

Pyright does seem to consider `tuple[ A | B]` vs `tuple[A] | tuple[B]` equivalent in a lot of cases, so I wonder if there's just a small blindspot.

https://pyright-play.net/?code=GYJw9gtgBALgngBwJYDsDmUkQWEMoAqiApgGoCGIAUDQMRTCrFUAmxwsI5KAzsLhAAUCAFywArggA2xANqp8AHwZSw5GAF0AlFAC0APgnS5CjVGUxJM2cFXqNIqlGdQQxSyBRQENKvUYozGwcMFy8-CAQAEzCYpbG8iia5kbWtmqaOgapJkkp6faOLq7u4p7evvQAgiBo4hDEeWAhJFAARPHWCikAcmCBGm1QAMbcKGD4AEbEUOQ8PEhogSywYN6U5A0wxCDtANbEcEPNsK0dVrnJFheyfQNtrOwM7sMAFoIAbuRSPGIsSMMYLJOpczNcEndiBoADSYJIwqAHOBxG7dZSQ7R6QwKIouNweLxfH6yJEaXxVKAAXkIJAoIEEbSqbS0VAAQlSaQgyJQGazmXQGExZmhyKgAPwSx4cYAvV4xIm-KD-QHAm4U5SshFgSYAK2IgIRSJRCXVUE1WUM2r1gNxznxZUJ3x4JMOZKAA

```python
from typing import TypeVar


# fine
def transform(p: tuple[int | float]) -> tuple[int] | tuple[float]:
    return p


# fine
def transform2(p: tuple[int] | tuple[float]) -> tuple[int | float]:
    return p


# Argument of type "tuple[int | None]" cannot be assigned to parameter "key" of type "tuple[int] | tuple[None]"
def fetch(vals: dict[tuple[int] | tuple[None], int], key: tuple[int | None]) -> int:
    return vals[key]


A = TypeVar("A")
B = TypeVar("B")

# fine again?
def fetch2(vals: dict[tuple[A | B], object], key: tuple[A | B]) -> object:
    return vals[key]
```

**Comment by kurtbrose** [comment]
Oh yea, this is really weird -- `None` is treated differently than `float`?

```python
# fine
def f(p: tuple[int | float]) -> tuple[int] | tuple[float]:
    return p


# Type "tuple[int | None]" is not assignable to return type "tuple[int] | tuple[None]"
def f2(p: tuple[int | None]) -> tuple[int] | tuple[None]:
    return p


# fine
def f3(vals: dict[tuple[int] | tuple[float], int], key: tuple[int | float]) -> int:
    return vals[key]


# Argument of type "tuple[int | None]" cannot be assigned to parameter "key" of type "tuple[int] | tuple[None]"
def f4(vals: dict[tuple[int] | tuple[None], int], key: tuple[int | None]) -> int:
    return vals[key]

```

https://pyright-play.net/?code=MQAgZglgdgpgUAExmcAKADgLhAFwK7oA2MA2tDiAD7iED2AhjgLoCUIAtAHy4HFlTMqPIqTB1GTTHBAyQAJxj45UEOjjrQAFQCe6GCABE%2BEfwrUAcrVhMDICAGcQUWhXr37EAOZR6AI2K4tPKKeMq4uvpGvKTkTELGfJbWBojI4ABMGNgJMQJCSTCsHNw5pnHUpQWS0rIKSipqGuDQ8EgoYADMqABu9IT22AgQAMY4JKWx8dEkYgzMADR2AkyLANYw2tnT5EKzEmxcSzhSssH1IL39JOvaTOpwoACCcp54ALYwebQoOBGGE3kLFZCrZhvQoM4KL59G4PN4YAhAqp6HJ6B8cDA5IYbrZvuE9P9tsspiYqik2uAACw9PoDEBDUbjImCCrTKqLWJrDZbEw7IHWA7ccgnWohMKXezXDZ3IA

**Comment by erictraut** [comment]
The type `int | float` is equivalent to the type `float` because `int` is a subtype of `float`. That means `tuple[int | float]` is assignable to `tuple[float]`. By contrast, `None` and `int` have no overlap, so `tuple[int | None]` is not assignable to either `tuple[int]` or `tuple[None]`.

**Comment by kurtbrose** [comment]
Thanks.  That would explain `int | float` vs `int | None`.  (Assuming this is a pyright thing that `int` and `float` have some kind of special handshake?)

What about the different behavior with `int | None` vs two `TypeVar`?  


```python
# Argument of type "tuple[int | None]" cannot be assigned to parameter "key" of type "tuple[int] | tuple[None]"
def fetch(vals: dict[tuple[int] | tuple[None], int], key: tuple[int | None]) -> int:
    return vals[key]


A = TypeVar("A")
B = TypeVar("B")

# fine again?
def fetch2(vals: dict[tuple[A | B], object], key: tuple[A | B]) -> object:
    return vals[key]
```

Surely there can't be any implied inheritance between `A` and `B` here right?  It seems that if `fetch2()` is type valid, `fetch()` should be as well, since it's more general?

**Comment by kurtbrose** [comment]
It also seems to be okay going in one direction but not the other

```python
# Type "tuple[int | None]" is not assignable to return type "tuple[int] | tuple[None]"
def tuple_or__or_tuple(p: tuple[int | None]) -> tuple[int] | tuple[None]:
    return p

# fine
def tuple_or__or_tuple__float(p: tuple[int | float]) -> tuple[int] | tuple[float]:
    return p

# fine
def or_tuple__tuple_or(p: tuple[int] | tuple[None]) -> tuple[int | None]:
    return p
```

https://pyright-play.net/?code=MQAgKgngDgpiBEAXArlANjA2gSwHaJAB8QA5Ae1xgF14RsBnEXMggQ3vuwHNdWAjDCERkQAJxgpRuIdDhJUGHPipEhCrOUo0AUABMYAMzXoYAfTKjT5yyhMAKKAC5jivAWKbqAShABaAHwuWG4qxLaKnlSO2iCxYhLIUiBQ2tqgBngweoZB1lYWpuFmpgZoZKyIDs5FSu4gpeWIVD4BQbWhbQ0VUTFx4pLSKWn1mdlGBUVWkxZVbSGqNZEtgTVuqpHRcfEDydpAA

**Comment by kurtbrose** [comment]
Is there some variance rule, `int` counts as `int | None` and `None` counts as `int | None'.
But `int | None` doesn't count as `int` or as `None`.

I can kind of see that if I squint.  I guess it's hard to predict which unions can be expanded and which can't.
